proposals/dom_create_port.md

```diff
@@ -0,0 +1,206 @@
+# Proposal: Inter (JS) World Communication
+
+**Summary**
+
+This introduces a special message port that can be passed and shared between different JS "worlds" (contexts) in the same document.
+
+**Document Metadata**
+
+**Author:** rdcronin
+
+**Sponsoring Browser:** Chrome
+
+**Contributors:** Rob--W, ...
+
+**Created:** 2024-06-24
+
+**Related Issues:** <TODO>
+
+## Motivation
+
+### Objective
+
+This allows the ability to establish a communication channel between different
+active JS worlds they may have.  This enables coordination between e.g. a main
+content script and any scripts that may inject in the main world, or between a
+content script and other user scripts.
+
+#### Use Cases
+
+The primary use case is coordinating work between different JS worlds in a way
+that is more difficult for any other script in the main world to intercept.
+This helps extensions avoid interacting with the existing page script, when
+doing so is undesirable.  This also allows executing code based on data within
+a content script without leaking the details of that content script to the main
+world.  (Though we don't consider the isolated world boundary a security
+boundary, it is useful isolation and can serve as a "first line of defense").
+
+### Known Consumers
+
+User script managers have expressed an interest in this API.  More broadly,
+this would help any extension that needs to communicate between the main world
+of a web page with a more trusted script.
+
+## Specification
+
+### Schema
+
+```
+declare namespace dom {
+  interface PortProperties {
+    world: ExecutionWorld;
+    worldId?: string;
+  };
+
+  interface MessagePort {
+    sendMessage(args?: any): void
+    onMessage(args?: any): void
+  };
+
+  export function createPort(
+      properties: PortProperties
+  ): MessagePort;
+}
+```
+
+### Behavior
+
+#### Creating a Port
+
+An extension can create a new message port using `dom.createPort()`.  This will
+create and return a new message port according to the specifies `properties`
+(see below).
+
+#### Port Properties and World ID
+
+A message port can only be used to communicate with a single other world.  This:
+* Avoids the "many-to-one" opener-listener behavior that existing message ports
+  (created by `runtime.connect()`) in extensions have.  This behavior has caused
+  increased complexity and developer confusion.  A given port will have only one
+  channel.
+* Protects against accidentally sharing the port with another world.
+
+The caller indicates which world it would like the port to be associated with
+by specifying the `world` property to the appropriate type of execution world
+and, if appropriate, the `worldId` (e.g. for specific user script worlds).
+
+#### Passing a Message Port
+
+A message port can be passed to another world (in order to establish a
+connection) by leveraging the new `dom.execute()` API.  `dom.execute()` will be
+expanded to have special serialization logic for the MessagePort type.
+
```

Comment:
I'd like to touch on the synchronocity aspect a bit more here (and I think return values are a natural byproduct of that).  I know we discussed this briefly before, but I'd like some more clarity on what this enables.  Given we have a synchronous dom.execute(), what the use cases in which synchronous message passing (as opposed to just FIFO) to a port is critical?  Is it for cases where the main world needs synchronous execution with the extension world?  If so, why?  Or for some other use case?

A not-necessarily-synchronous design leaves this more open -- we could potentially have a single world blocked while another is unblocked.  I'm honestly not sure if this exists today in Chromium (I'm not sure about other browsers), but I could see definite situations in which it could (e.g., we might pause the main world, but allow extensions to continue running).  If we guarantee this synchronocity of message passing, that means we also need to block execution in the extension world on any execution that may be happening in the main world (preventing extensions from doing work they may otherwise be able to).

Reply:
Generally, synchronous message passing enables the already injected main world script to reach back to the isolated world to look up relevant state that the script in the main world needs.

Other extension authors have already commented on some use cases at https://github.com/w3c/webextensions/issues/284

As an example of a concrete use case from personal experience: I have an extension called Don't Track Me Google, which is just a content script that ensures that <a href="//google.com/url=realurl">` becomes `<a href="realurl">`. Among other things, the extension intercepts the setter of `HTMLAnchorElement.prototype.href`, to hook in on calls from the web page. It sanitizes it and calls the original setter.

To minimize interference from the web page, the sanitization logic should run in an isolated world. And because the original caller expects a synchronous API, the messaging API has to be synchronous.

This specific use case also shows why sometimes it is desirable to be able to receive DOM objects, so that the page and caller can pass the DOM elements (that they already share!) without requiring appending to the DOM.

Note: this extension was originally a MV2 extension. To convert to MV3, I had to duplicate the sanitization logic in the main world and use a shared DOM element for communicating specific state (preferences from the options page) from the content script to the main world script: https://github.com/Rob--W/dont-track-me-google/blob/a73ab8b3bfa9794c75cb4e111c6424e31e6c019e/main_world_script.js. The shared DOM element can be detected by the web page. In my case it's a reasonable trade-off, especially because I don't expect Google web properties to intentionally break the extension. This assumption is not necessarily true for extensions that are privacy-focused and designed for the whole web.

> A not-necessarily-synchronous design leaves this more open -- we could potentially have a single world blocked while another is unblocked. I'm honestly not sure if this exists today in Chromium (I'm not sure about other browsers), but I could see definite situations in which it could (e.g., we might pause the main world, but allow extensions to continue running). If we guarantee this synchronocity of message passing, that means we also need to block execution in the extension world on any execution that may be happening in the main world (preventing extensions from doing work they may otherwise be able to).

There is already a synchronous communication between worlds - they share the same DOM, and therefore DOM events.

Reply:
> Given we have a synchronous dom.execute()

Abusing it to pass messages synchronously would be very inefficient as running code requires a lot more internal work on top passing the arguments. Userscripts may exchange thousands of messages between the worlds.

> use cases in which synchronous message passing [...] is critical

It allows implementing simple super-performant API via object getters exposed to the userscript or main worlds that receive the actual info from the secure world synchronously.

> we could potentially have a single world blocked while another is unblocked

They share the same JS/DOM thread, so what you describe is impossible.

> If we guarantee this synchronocity of message passing, that means we also need to block execution in the extension world on any execution that may be happening in the main world (preventing extensions from doing work they may otherwise be able to).

No, we just abandon the idea of MessagePort altogether and use CustomEvent mechanism internally, which already does it synchronously and which is what userscript extensions have been always using. The side benefit would be the API may be simplified as there's no need for a port.